bzzeth/peer.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package bzzeth
+
+import (
+	"math/rand"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+// Peer extends p2p/protocols Peer and represents a conrete protocol connection
+type Peer struct {
+	*protocols.Peer            // embeds protocols.Peer
+	serveHeaders    bool       // if the remote serves headers
+	requests        *requests  // per-peer pool of open requests
+	logger          log.Logger // custom logger for peer
+}
+
+// NewPeer is the constructor for Peer
+func NewPeer(peer *protocols.Peer) *Peer {
+	return &Peer{
+		Peer:     peer,
+		requests: newRequests(),
+		logger:   log.New("peer", peer.ID()),
+	}
+}
+
+// peers represents the bzzeth specific peer pool
+type peers struct {
+	mtx   sync.RWMutex
+	peers map[enode.ID]*Peer
+}
+
+func newPeers() *peers {
+	return &peers{peers: make(map[enode.ID]*Peer)}
+}
+
+func (p *peers) get(id enode.ID) *Peer {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+	return p.peers[id]
+}
+
+func (p *peers) add(peer *Peer) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	p.peers[peer.ID()] = peer
+}
+
+func (p *peers) remove(peer *Peer) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	delete(p.peers, peer.ID())
+}
+
+// getEthPeer finds a peer that serves headers and calls the function argument on this peer
+// TODO: implement load balancing of requests in case of multiple peers
+func (p *peers) getEth() (peer *Peer) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+	for _, peer = range p.peers {
```

Comment:
Nice but unusual, at least for me. I would expect something like this in other languages, but here, I would use := operator in for range and return peer instead to break. it is just because I needed to think a few seconds more to be sure that final return is not always nil. If someone changes for range loop to use := operator for any reason and leave the break, the function scope peer variable will always be returned as nil. Just a suggestion. This implementation is functional as it is.

Reply:
Got it. Done